docs/configuration.rst
.. _configuration:

Configuration
=============

Upon first import, ``osyris`` will create a configuration file located in
``/home/user/.osyris/config_osyris.py``.
This will then be loaded by ``osyris``, and allows easy configuration for the
default colormap, spatial scale, time unit etc.

Default values
--------------

- ``"nout"`` : 1 : The output number to load.
- ``"lmax"`` : 0 : The maximum level of cells that will be loaded
  (0 = no limit).
- ``"center"`` : None : The center around which the grid will be centered.
- ``"dx"`` : 0.0 : The ``x`` extent of the computational domain to be loaded.
- ``"dy"`` : 0.0 : The ``y`` extent of the computational domain to be loaded.
- ``"dz"`` : 0.0 : The ``z`` extent of the computational domain to be loaded.
- ``"scale"`` : "au" : The scale for the ``x, y, z`` coordinates.
- ``"time_unit"`` : "kyr" : The time unit (for labels, titles, etc.).
- ``"verbose"`` : False : Print data information when loaded if ``True``.
- ``"path"`` : "" : Path to the directory where to read outputs from,
  if different from the current directory.
- ``"variables"`` : [] : List of variables to be loaded.
- ``"colormap"`` : "viridis" : Color map for slices, images, etc.


Physical constants
------------------
- ``"cm"`` : 1.0 : Centimeter.
- ``"au"`` : 1.495980e+13 : Astronomical unit.
- ``"pc"`` : 3.085678e+18 : Parsec.
- ``"s"`` : 1.0 : Second.
- ``"yr"`` : 365.25*86400.0 : Year.
- ``"kyr"`` : 365.25*86400.0*1000.0 : Kilo-year.
- ``"msun"`` : 1.9889e33 : Solar mass.
- ``"a_r"`` : 7.56591469318689378e-015 : Radiation constant.
- ``"c"`` : 2.9979250e+10 : Speed of light.


Additional colormaps
--------------------

.. raw:: html

   <ul>
   <li>"osyris" :
   <svg width="414" height="24" >
     <linearGradient id="osyris" x1="0" x2="1" y1="0" y2="0">
       <stop style="stop-color:#2b3c4e;stop-opacity:1" offset="0%"/>
       <stop style="stop-color:#249593;stop-opacity:1" offset="33.3%"/>
       <stop style="stop-color:#db6a6c;stop-opacity:1" offset="66.7%"/>
       <stop style="stop-color:#ffffff;stop-opacity:1" offset="100%"/>
     </linearGradient>
     <rect x="10" y="2" width="400" height="20" style="fill:url(#osyris);stroke-width:2;stroke:#000000"/>
   </svg>
   <li>"osyris2" :
   <svg width="404" height="24" >
     <linearGradient id="osyris2" x1="0" x2="1" y1="0" y2="0">
       <stop style="stop-color:#2b3c4e;stop-opacity:1" offset="0%"/>
       <stop style="stop-color:#249593;stop-opacity:1" offset="25%"/>
       <stop style="stop-color:#ffffff;stop-opacity:1" offset="50%"/>
       <stop style="stop-color:#db6a6c;stop-opacity:1" offset="75%"/>
       <stop style="stop-color:#9e4d4e;stop-opacity:1" offset="100%"/>
     </linearGradient>
     <rect x="2" y="2" width="400" height="20" style="fill:url(#osyris2);stroke-width:2;stroke:#000000"/>
   </svg>
   <li>"osyris3" :
   <svg width="404" height="24" >
     <linearGradient id="osyris3" x1="0" x2="1" y1="0" y2="0">
       <stop style="stop-color:#3d3d6b;stop-opacity:1" offset="0%"/>
       <stop style="stop-color:#2a7b9b;stop-opacity:1" offset="10%"/>
       <stop style="stop-color:#00baad;stop-opacity:1" offset="20%"/>
       <stop style="stop-color:#57c785;stop-opacity:1" offset="30%"/>
       <stop style="stop-color:#add45c;stop-opacity:1" offset="40%"/>
       <stop style="stop-color:#ffc300;stop-opacity:1" offset="50%"/>
       <stop style="stop-color:#ff8d1a;stop-opacity:1" offset="60%"/>
       <stop style="stop-color:#ff5733;stop-opacity:1" offset="70%"/>
       <stop style="stop-color:#c70039;stop-opacity:1" offset="80%"/>
       <stop style="stop-color:#900c3f;stop-opacity:1" offset="90%"/>
       <stop style="stop-color:#511849;stop-opacity:1" offset="100%"/>
     </linearGradient>
     <rect x="2" y="2" width="400" height="20" style="fill:url(#osyris3);stroke-width:2;stroke:#000000"/>
   </svg>
   <li>"osyris4" :
   <svg width="404" height="24" >
     <linearGradient id="osyris4" x1="0" x2="1" y1="0" y2="0">
       <stop style="stop-color:#000000;stop-opacity:1" offset="0%"/>
       <stop style="stop-color:#ff5b00;stop-opacity:1" offset="14.3%"/>
       <stop style="stop-color:#ffff00;stop-opacity:1" offset="28.6%"/>
       <stop style="stop-color:#00ff00;stop-opacity:1" offset="42.9%"/>
       <stop style="stop-color:#2bc184;stop-opacity:1" offset="57.1%"/>
       <stop style="stop-color:#3d3d6b;stop-opacity:1" offset="71.4%"/>
       <stop style="stop-color:#ffffff;stop-opacity:1" offset="85.7%"/>
       <stop style="stop-color:#0000ff;stop-opacity:1" offset="100%"/>
     </linearGradient>
     <rect x="2" y="2" width="400" height="20" style="fill:url(#osyris4);stroke-width:2;stroke:#000000"/>
   </svg>
   <li>"osyris5" :
   <svg width="404" height="24" >
     <linearGradient id="osyris5" x1="0" x2="1" y1="0" y2="0">
       <stop style="stop-color:#0000ff;stop-opacity:1" offset="0%"/>
       <stop style="stop-color:#00ffff;stop-opacity:1" offset="16.7%"/>
       <stop style="stop-color:#00ff00;stop-opacity:1" offset="33.3%"/>
       <stop style="stop-color:#ffff00;stop-opacity:1" offset="50%"/>
       <stop style="stop-color:#ff0000;stop-opacity:1" offset="66.7%"/>
       <stop style="stop-color:#000000;stop-opacity:1" offset="83.3%"/>
       <stop style="stop-color:#ffffff;stop-opacity:1" offset="100%"/>
     </linearGradient>
     <rect x="2" y="2" width="400" height="20" style="fill:url(#osyris5);stroke-width:2;stroke:#000000"/>
   </svg>
   </ul>

Additional variables
--------------------

Define variables, using either operations on existing variables or new values,
that will be created every time an output is loaded.

For instance, to create the logarithm of gas density, we use

.. code-block:: python

   holder.new_field(name="log_rho", operation="np.log10(density)",
                    unit="g/cm3", label="log(Density)", verbose=False)


src/osyris/config.py
# SPDX-License-Identifier: GPL-3.0-or-later
# Copyright (c) 2019 Osyris contributors (https://github.com/nvaytet/osyris)
# @author Neil Vaytet

from matplotlib.colors import LinearSegmentedColormap
import matplotlib.pyplot as plt
import numpy as np

#==============================================================================
# Define default values so that you don't have to specify them every time.
#==============================================================================
default_values = {
    "nout"        : 1    ,
    "lmax"        : 0    ,
    "center"      : None ,
    "dx"          : 0.0  ,
    "dy"          : 0.0  ,
    "dz"          : 0.0  ,
    "scale"       : "au" ,
    "time_unit"   : "kyr",
    "verbose"     : False,
    "path"        : ""   ,
    "variables"   : []   ,
    "colormap"    : "viridis"
}

#==============================================================================
# Define default units for some selected variables. The syntax is:
# "my_variable_name" : ["combination of unit_d, unit_l, unit_t" , "string to be displayed on grid axes"]
#==============================================================================
default_units = {
    ## Example for internal energy
    #"passive_scalar_4" : ["unit_d*((unit_l/unit_t)**2)" , "erg/cm3"],
}

#==============================================================================
# Physical constants
#==============================================================================
constants = {
    "cm"  : 1.0,
    "au"  : 1.495980e+13,
    "pc"  : 3.085678e+18,
    "s"   : 1.0,
    "yr"  : 365.25*86400.0,
    "kyr" : 365.25*86400.0*1000.0,
    "msun": 1.9889e33,
    "a_r" : 7.56591469318689378e-015,
    "c"   : 2.9979250e+10
}

#==============================================================================
# Custom colormaps
#==============================================================================
cmaps = {
    "osyris"  : ["#2b3c4e", "#249593", "#db6a6c", "#ffffff"],
    "osyris2" : ["#2b3c4e", "#249593", "#ffffff", "#db6a6c", "#9e4d4e"],
    "osyris3" : ["#3d3d6b", "#2a7b9b", "#00baad", "#57c785", "#add45c",
                 "#ffc300", "#ff8d1a", "#ff5733", "#c70039", "#900c3f",
                 "#511849"],
    "osyris4" : ["#000000", "#ff5b00", "#ffff00", "#00ff00", "#2bc184",
                 "#3d3d6b", "#ffffff", "#0000ff"],
    "osyris5" : ["#0000ff", "#00ffff", "#00ff00", "#ffff00", "#ff0000",
                 "#000000", "#ffffff"]
}

for key in cmaps.keys():
    cmap = LinearSegmentedColormap.from_list(key, cmaps[key])
    cmap_r = LinearSegmentedColormap.from_list(key+"_r", cmaps[key][::-1])
    plt.register_cmap(cmap=cmap)
    plt.register_cmap(cmap=cmap_r)


def additional_variables(holder):
    """
    Here are some additional variables that are to be computed every time data
    is loaded.
    """

    # Velocity field (in case conservative variables are dumped)
    holder.new_field(name="velocity", operation="momentum/density",
                     unit="cm/s", label="velocity", verbose=False)

    # Magnetic field
    holder.new_field(name="B", operation="0.5*(B_left+B_right)", unit="G",
                     label="B", verbose=False)

    # Mass
    holder.new_field(name="mass",
                     operation="density*((dx*"+str(constants[holder.info["scale"]])+")**3)/"+str(constants["msun"]),
                     unit="Msun", label="Mass", verbose=False)

    # Commonly used log quantities
    holder.new_field(name="log_rho", operation="np.log10(density)",
                     unit="g/cm3", label="log(Density)", verbose=False)
    holder.new_field(name="log_T", operation="np.log10(temperature)", unit="K",
                     label="log(T)", verbose=False)
    holder.new_field(name="log_B", operation="np.log10(B.values)", unit="G",
                     label="log(B)", verbose=False)
    holder.new_field(name="log_m", operation="np.log10(mass)", unit="Msun",
                     label="log(Mass)", verbose=False)

    # Photon density (in case conservative variables are not dumped)
    if hasattr(holder,"photon_flux_density_1"):
        for igrp in range(holder.info_rt["nGroups"]):
            holder.new_field(name="photon_density_"+str(igrp+1),
                             operation="photon_flux_density_"+str(igrp+1)+"/("+str(constants["c"]*holder.info_rt["rt_c_frac"])+")",
                             unit="photons/cm3",
                             label="photon density "+str(igrp+1),
                             verbose=False)

    #========================== ADD YOUR VARIABLES HERE ============================

    return
